### PR TITLE
enh(tcp): Move to a totaly asynchronous read

### PR DESCRIPTION
**Pull request by SylvestreG**
architecture for asio code. This change will :
- increase perfs
- allow us to detect connection lost on adaptor and connector.

# Pull Request Template

## Description

Please include a short resume of the changes and what is the purpose of PR. Any relevant information should be added to help:
* **QA Team** (Quality Assurance) with tests.
* **reviewers** to understand what are the stakes of the pull request.

**Fixes** #4628

## Type of change

- [X] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 18.10.x
- [ ] 19.04.x
- [ ] 19.10.x
- [X] 20.04.x (master)

<h2> How this pull request can be tested ? </h2>

Simon will test it

#### Community contributors & Centreon team

- [X] I followed the **coding style guidelines** provided by Centreon
- [ ] I have made corresponding changes to the **documentation**.
- [X] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [X] I have made sure that the **unit tests** related to the story are successful.
- [X] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: tcp/src/acceptor.cc [hunk 1]
```diff
@@ -60,6 +62,19 @@ void acceptor::add_child(std::string const& child) {
  */
 void acceptor::listen_on(unsigned short port) {
   _port = port;
+  _acceptor.reset(nullptr);
```
Comment (bouda1) [inline tcp/src/acceptor.cc]: I think _acceptor.reset(); works too

File: tcp/src/stream.cc [hunk 2]
```diff
@@ -70,13 +69,11 @@ stream::stream(asio::io_context& ctx,
 /**
  *  Destructor.
  */
+#include <iostream>
```
Comment (bouda1) [inline tcp/src/stream.cc]: Do you really need this header?

File: tcp/inc/com/centreon/broker/tcp/tcp_async.hh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  ~tcp_async();
+  void _async_job();
+  void _async_read_cb(asio::ip::tcp::socket &socket,
+                      int fd,
+                      std::error_code const& ec,
+                      std::size_t bytes);
+
+ public:
+  void register_socket(asio::ip::tcp::socket &socket);
+  void unregister_socket(std::shared_ptr<asio::ip::tcp::socket> socket);
```
[the remark this answers is not in the document]
Comment (SylvestreG) [reply tcp/inc/com/centreon/broker/tcp/tcp_async.hh]: unregister specifically needs a shared_ptr, I will add a comment to explain this behaviour.

File: tcp/src/stream.cc (hunk 2 above)
Comment (SylvestreG) [reply tcp/src/stream.cc]: it is a leftover.

**Pull request merged**